### ISSUE TITLE
Populating a concretization of a Generic Dictionary does not work in XAML

### ISSUE THREAD

**Issue by npolyak** [AvaloniaUI/Avalonia#7065]
This happens in the latest version 0.10.10
I have a simple sample that reproduces this behavior under [NP.Demos.XamlGenericsBugSample](https://github.com/npolyak/NP.Avalonia.Demos/tree/main/NP.Demos.XamlSamples/NP.Demos.XamlGenericsBugSample). 
For example I created a concretization ```public class IntToStrDictionary : Dictionary<int, string>{}``` of a generic ```Dictionary<TKey, TValue>```.
I am trying to populate it within MainWindow.axaml file with Key=1 mapped into "str 1" string:
 

    <Grid.DataContext>
           
           <local:IntToStrDictionary>
                 <x:String x:Key="1">str 1</x:String>
          </local:IntToStrDictionary>
     </Grid.DataContext>

The code above compiles, but trying to run it will through a ```System.NullReferenceException``` at the line ```<local:IntToStrDictionary>``` within XAML file. Here is the stack trace:

       at System.Runtime.CompilerServices.CastHelpers.ChkCastClass(Void* toTypeHnd, Object obj)
       at NP.Demos.XamlGenericsBugSample.MainWindow.!XamlIlPopulate(IServiceProvider , MainWindow )

If I comment out the line that populates the dictionary ```  <!--<x:String x:Key="1">str 1</x:String>-->  ```, no exception is thrown - application starts fine.

Using the generic arguments also works:

    <collections:Dictionary x:TypeArguments="x:Int32, x:String">
        <x:String x:Key="1">str 1</x:String>
    </collections:Dictionary>

A concretization of a List, eg. ```public class ListOfStrings : List<string> {}``` also works fine. This is something specific to a concretization of a generic ```Dictionary<TKey, TValue>```.

**Comment by robloo** [comment]
Starting a name with a digit is illegal in c#. I imagine that is enforced in XAML as well. Your first issue doesn't seem like an issue to be though it likely should throw a compile error.

https://docs.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/identifier-names

**Comment by maxkatz6** [comment]
I don't think key matters here. It can be any string or int or any other object.

**Comment by npolyak** [comment]
Yes, the key type can be anything. Both examples that break should work. They would work in WPF also.
